File: src/algorithm_quiz/java/leetcode/medium/ProductOfArrayExceptSelf.java
package algorithm_quiz.java.leetcode.medium;

public class ProductOfArrayExceptSelf {
     /*
    238. Product of Array Except Self
     */

    public int[] productExceptSelf(int[] nums) {
        int[] result = new int[nums.length];
        int a = 1;
        for (int i = 0; i < nums.length; i++) {
            result[i] = a;
            a *= nums[i];
        }
        a = 1;
        for (int i = nums.length-1; 0 <= i; i--) {
            result[i] *= a;
            a *= nums[i];
        }
        return result;
    }
}

File: src/algorithm/kotlin/leetcode/medium/ProductOfArrayExceptSelf.kt
package algorithm.kotlin.leetcode.medium

class ProductOfArrayExceptSelf {
    /*
    238. Product of Array Except Self
     */

    fun productExceptSelf(nums: IntArray): IntArray {
        val result = IntArray(nums.size)
        var a = 1
        for (i in nums.indices) {
            result[i] = a
            a *= nums[i]
        }
        a = 1
        for (i in nums.indices.reversed()) {
            result[i] *= a
            a *= nums[i]
        }
        return result
    }
}